### PR TITLE
L1 Syncing Thread

### LINKED ISSUE #1177: create a l1_event_listener lib crate (or some other name)

### PR DESCRIPTION
**Pull request by bidzyyys**
Closes #1177 
A rewrite of https://github.com/keep-starknet-strange/madara/pull/1205

### CODE REVIEW

File: crates/client/l1-messages/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+use std::fs::File;
+use std::path::PathBuf;
+
+use ethers::types::Address;
+
+use crate::error::L1MessagesConfigError;
+
+pub const DEFAULT_ETHEREUM_NODE: &str = "127.0.0.1:8545";
```
Comment (bidzyyys) [inline crates/client/l1-messages/src/config.rs:8]: To be removed

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/l1-messages/src/worker.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                log::info!(
+                    "⟠ L1 Message from block: {:?}, transaction_hash: {:?}, log_index: {:?} submitted, transaction \
+                     hash on L2: {:?}",
+                    meta.block_number,
+                    meta.transaction_hash,
+                    meta.log_index,
+                    tx_hash
+                );
+            }
+            Err(e) => {
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply crates/client/l1-messages/src/worker.rs:74]: We can think about it in the next PR if needed.

File: crates/node/src/commands/run.rs [hunk 3]
```diff
@@ -2,15 +2,48 @@ use std::path::PathBuf;
...
+    pub l1_contract_address: String,
+}
+
+#[derive(Debug, Clone, clap::Args)]
+pub struct L1Messages {
+    /// Path to configuration file for Ethereum Core Contract Events Listener
+    #[clap(
+        long,
+        conflicts_with="provider_url",
+        conflicts_with="l1_contract_address",
```
Comment (tdelabro) [inline crates/node/src/commands/run.rs:38]: I think you can make it conflict with the ArgGroup directly

File: crates/node/src/commands/run.rs [hunk 4]
```diff
@@ -2,15 +2,48 @@ use std::path::PathBuf;
...
+        requires="l1_contract_address",
+        value_hint=clap::ValueHint::Url,
+    )]
+    pub provider_url: String,
+
+    /// L1 Contract Address
+    #[clap(
+        long,
+        requires="provider_url",
+        value_hint=clap::ValueHint::Other,
```
Comment (tdelabro) [inline crates/node/src/commands/run.rs:27]: No need for those two requires. If they are not Option<T>, I think that they are both required

File: crates/node/src/commands/run.rs [hunk 5]
```diff
@@ -2,15 +2,48 @@ use std::path::PathBuf;
...
+    #[clap(
+        long,
+        conflicts_with="provider_url",
+        conflicts_with="l1_contract_address",
+        value_hint=clap::ValueHint::FilePath,
+    )]
+    pub l1_messages_config: Option<PathBuf>,
+
+    #[clap(flatten)]
+    pub config_params: Option<L1MessagesParams>,
```
Comment (tdelabro) [inline crates/node/src/commands/run.rs:44]: Here you can probably state that this one conflict with `l1_messages_config`

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/node/src/commands/run.rs (hunk 5 above)
Comment (bidzyyys) [reply crates/node/src/commands/run.rs:44]:
```sh
error: methods are not allowed for flattened entry
  --> crates/node/src/commands/run.rs:41:12
   |
41 |     #[clap(flatten, conflicts_with="l1_messages_config")]
```

File: crates/node/src/commands/run.rs (hunk 3 above)
Comment (bidzyyys) [reply crates/node/src/commands/run.rs:38]:
```
thread 'main' panicked at /Users/danielbigos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.8/src/builder/debug_asserts.rs:214:13:
Command madara: Argument or group 'config_params' specified in 'conflicts_with*' for 'l1_messages_config' does not exist
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: crates/node/src/commands/run.rs (hunk 5 above)
Comment (tdelabro) [reply crates/node/src/commands/run.rs:44]: and if you don't flatten?

Comment (tdelabro) [reply crates/node/src/commands/run.rs:42]: what we want express is that only one of the two should be used.
Maybe the good way is to put them in an arg group and there is some "only one in those" syntax

Comment (bidzyyys) [reply crates/node/src/commands/run.rs:42]: without `flatten`
```sh
error[E0599]: the method `value_parser` exists for reference `&&&&&&_AutoValueParser<L1MessagesParams>`, but its trait bounds were not satisfied
    --> crates/node/src/commands/run.rs:41:5
     |
14   |   pub struct L1MessagesParams {
     |   ---------------------------
     |   |
     |   doesn't satisfy `L1MessagesParams: From<&'s std::ffi::OsStr>`
     |   doesn't satisfy `L1MessagesParams: From<&'s str>`
     |   doesn't satisfy `L1MessagesParams: From<OsString>`
     |   doesn't satisfy `L1MessagesParams: From<std::string::String>`
     |   doesn't satisfy `L1MessagesParams: FromStr`
     |   doesn't satisfy `L1MessagesParams: ValueEnum`
     |   doesn't satisfy `L1MessagesParams: ValueParserFactory`
...
41   | /     #[clap(conflicts_with="l1_messages_config")]
42   | |     pub config_params: Option<L1MessagesParams>,
     | |_______________________________________________^ method cannot be called on `&&&&&&_AutoValueParser<L1MessagesParams>` due to unsatisfied trait bounds
     |
    ::: /Users/danielbigos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.8/src/builder/value_parser.rs:2462:1
     |
2462 |   pub struct _AutoValueParser<T>(std::marker::PhantomData<T>);
     |   ------------------------------ doesn't satisfy `_: _ValueParserViaParse`
     |
     = note: the following trait bounds were not satisfied:
             `L1MessagesParams: ValueEnum`
             which is required by `&&&&&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaValueEnum`
             `L1MessagesParams: ValueParserFactory`
             which is required by `&&&&&&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaFactory`
             `L1MessagesParams: From<OsString>`
             which is required by `&&&&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaFromOsString`
             `L1MessagesParams: From<&'s std::ffi::OsStr>`
             which is required by `&&&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaFromOsStr`
             `L1MessagesParams: From<std::string::String>`
             which is required by `&&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaFromString`
             `L1MessagesParams: From<&'s str>`
             which is required by `&_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaFromStr`
             `L1MessagesParams: FromStr`
             which is required by `_AutoValueParser<L1MessagesParams>: clap::builder::via_prelude::_ValueParserViaParse`
note: the traits `ValueEnum`, `ValueParserFactory`, `From`,  and `FromStr` must be implemented
    --> /Users/danielbigos/.rustup/toolchains/nightly-2023-08-24-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/str/traits.rs:603:1
     |
603  | pub trait FromStr: Sized {
     | ^^^^^^^^^^^^^^^^^^^^^^^^
     |
    ::: /Users/danielbigos/.rustup/toolchains/nightly-2023-08-24-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:538:1
     |
538  | pub trait From<T>: Sized {
     | ^^^^^^^^^^^^^^^^^^^^^^^^
     |
    ::: /Users/danielbigos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.8/src/derive.rs:264:1
     |
264  | pub trait ValueEnum: Sized + Clone {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
    ::: /Users/danielbigos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.8/src/builder/value_parser.rs:2312:1
     |
2312 | pub trait ValueParserFactory {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: this error originates in the macro `clap::value_parser` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
error: could not compile `madara` (bin "madara") due to 2 previous errors
```

File: crates/node/src/commands/run.rs (hunk 3 above)
Comment (bidzyyys) [reply crates/node/src/commands/run.rs:36]: Cannot use `exclusive=true`
```sh
Running `target/debug/madara --dev --l1-messages-config ./examples/messaging/l1-messaging-config.json`
error: the argument '--l1-messages-config <L1_MESSAGES_CONFIG>' cannot be used with one or more of the other specified arguments

Usage: madara [OPTIONS] --provider-url <PROVIDER_URL> --l1-contract-address <L1_CONTRACT_ADDRESS>
       madara <COMMAND>
```

**Pull request merged**